### PR TITLE
Add virtual destructor for abstract classes

### PR DESCRIPTION
**Pull request by sparfenyuk**
Avoid undefined behavior by setting virtual destructor for base abstract classes.

### CODE REVIEW

File: src/QtImageReader.cpp [hunk 1]
```diff
@@ -209,7 +209,7 @@ std::shared_ptr<Frame> QtImageReader::GetFrame(int64_t requested_frame)
 	}
 
 	// Scale image smaller (or use a previous scaled image)
-	if (!cached_image || (cached_image && cached_image->width() != max_width || cached_image->height() != max_height)) {
```
Comment (jonoomph) [inline src/QtImageReader.cpp]: If cached_image is NULL, will this still work? Or do we still need `cached_image && cached_image->...`

**Comment by jonoomph** [comment]
Thanks @sparfenyuk , just a quick question for you about one of the changes.

File: src/QtImageReader.cpp (hunk 1 above)
Comment (sparfenyuk) [reply src/QtImageReader.cpp]: In case if `cached_image` is NULL then the second part of the logical expression will not be evaluated.

**Comment by sparfenyuk** [comment]
@jonoomph If there are no questions anymore let's merge then :)

File: src/AudioResampler.cpp [hunk 2]
```diff
@@ -74,9 +74,9 @@ AudioResampler::~AudioResampler()
 void AudioResampler::SetBuffer(AudioSampleBuffer *new_buffer, double sample_rate, double new_sample_rate)
 {
 	if (sample_rate <= 0)
-		sample_rate == 44100;
+		sample_rate = 44100;
 	if (new_sample_rate <= 0)
-		new_sample_rate == 44100;
+		new_sample_rate = 44100;
```
Comment (ferdnyc) [inline src/AudioResampler.cpp]: I can't believe `g++` doesn't catch this and throw up _some_ sort of warning... makes me wonder if we're missing some compiler flags, and don't have all of the sanity-checking warnings enabled that we should?

Comment (ferdnyc) [reply src/AudioResampler.cpp]: Hah! Turns out `-Wall` totally _would_ have caught that... if we were building with `-Wall`. Which we're not.

```console
$ cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_CXX_FLAGS=-Wall ..
$ # [...]
$ make VERBOSE=1
$ # [...]
[ 10%] Building CXX object src/CMakeFiles/openshot.dir/AudioResampler.cpp.o
cd /home/ferd/rpmbuild/REPOS/libopenshot/build/src && /usr/lib64/ccache/c++  -DMAGICKCORE_HDRI_ENABLE=0 -DMAGICKCORE_QUANTUM_DEPTH=16 -DOPENSHOT_IMAGEMAGICK_COMPATIBILITY=0 -DQT_CORE_LIB -DQT_GUI_LIB -DQT_MULTIMEDIAWIDGETS_LIB -DQT_MULTIMEDIA_LIB -DQT_NETWORK_LIB -DQT_WIDGETS_LIB -DUSE_IMAGEMAGICK=1 -Dopenshot_EXPORTS -I/usr/include/ImageMagick-6 -I/usr/include/ffmpeg -I/usr/include/libopenshot-audio -isystem /usr/include/qt5 -isystem /usr/include/qt5/QtWidgets -isystem /usr/include/qt5/QtGui -isystem /usr/include/qt5/QtCore -isystem /usr/lib64/qt5/mkspecs/linux-g++ -isystem /usr/include/qt5/QtMultimedia -isystem /usr/include/qt5/QtNetwork -isystem /usr/include/qt5/QtMultimediaWidgets  -Wall           -g -ggdb  -fopenmp  -g -fPIC   -fPIC -std=c++11 -o CMakeFiles/openshot.dir/AudioResampler.cpp.o -c /home/ferd/rpmbuild/REPOS/libopenshot/src/AudioResampler.cpp
/home/ferd/rpmbuild/REPOS/libopenshot/src/AudioResampler.cpp: In member function ‘void openshot::AudioResampler::SetBuffer(juce::AudioSampleBuffer*, double, double)’:
/home/ferd/rpmbuild/REPOS/libopenshot/src/AudioResampler.cpp:77:15: warning: statement has no effect [-Wunused-value]
   77 |   sample_rate == 44100;
      |   ~~~~~~~~~~~~^~~~~~~~
/home/ferd/rpmbuild/REPOS/libopenshot/src/AudioResampler.cpp:79:19: warning: statement has no effect [-Wunused-value]
   79 |   new_sample_rate == 44100;
      |   ~~~~~~~~~~~~~~~~^~~~~~~~
```

...Why aren't we building with `-Wall` (at the very _least_)?

Comment (jonoomph) [reply src/AudioResampler.cpp]: Sounds like we need Travis and GitLab to be building with `-DCMAKE_CXX_FLAGS=-Wall`

**Comment by jonoomph** [comment]
LGTM! Thanks

**Comment by jonoomph** [comment]
Oops, this one is breaking our GitLab builders. Output below:

```
  1%] Generating __/include/Qt/moc_PlayerPrivate.cpp
/home/gitlab-runner/builds/5cd61c66/0/OpenShot/libopenshot/include/Qt/PlayerPrivate.h:0: Note: No relevant classes found. No output generated.
[  2%] Generating __/include/Qt/moc_AudioPlaybackThread.cpp
/home/gitlab-runner/builds/5cd61c66/0/OpenShot/libopenshot/include/Qt/AudioPlaybackThread.h:0: Note: No relevant classes found. No output generated.
[  3%] Generating __/include/Qt/moc_VideoCacheThread.cpp
/home/gitlab-runner/builds/5cd61c66/0/OpenShot/libopenshot/include/Qt/VideoCacheThread.h:0: Note: No relevant classes found. No output generated.
[  4%] Generating __/include/Qt/moc_VideoPlaybackThread.cpp
/home/gitlab-runner/builds/5cd61c66/0/OpenShot/libopenshot/include/Qt/VideoPlaybackThread.h:0: Note: No relevant classes found. No output generated.
[  5%] Generating __/include/Qt/moc_PlayerDemo.cpp
[  6%] Generating __/include/Qt/moc_VideoRenderWidget.cpp
[  7%] Generating __/include/Qt/moc_VideoRenderer.cpp
Scanning dependencies of target openshot
[  8%] Building CXX object src/CMakeFiles/openshot.dir/AudioBufferSource.cpp.o
[  9%] Building CXX object src/CMakeFiles/openshot.dir/AudioReaderSource.cpp.o
In file included from /home/gitlab-runner/builds/5cd61c66/0/OpenShot/libopenshot/src/../include/CacheMemory.h:37:0,
                 from /home/gitlab-runner/builds/5cd61c66/0/OpenShot/libopenshot/src/../include/ReaderBase.h:39,
                 from /home/gitlab-runner/builds/5cd61c66/0/OpenShot/libopenshot/src/../include/AudioReaderSource.h:43,
                 from /home/gitlab-runner/builds/5cd61c66/0/OpenShot/libopenshot/src/AudioReaderSource.cpp:31:
/home/gitlab-runner/builds/5cd61c66/0/OpenShot/libopenshot/src/../include/CacheBase.h:117:11: error: ‘virtual openshot::CacheBase::~CacheBase()’ cannot be overloaded
   virtual ~CacheBase() = default;
           ^
/home/gitlab-runner/builds/5cd61c66/0/OpenShot/libopenshot/src/../include/CacheBase.h:66:11: error: with ‘virtual openshot::CacheBase::~CacheBase()’
   virtual ~CacheBase();
           ^
In file included from /home/gitlab-runner/builds/5cd61c66/0/OpenShot/libopenshot/src/../include/ReaderBase.h:41:0,
                 from /home/gitlab-runner/builds/5cd61c66/0/OpenShot/libopenshot/src/../include/AudioReaderSource.h:43,
                 from /home/gitlab-runner/builds/5cd61c66/0/OpenShot/libopenshot/src/AudioReaderSource.cpp:31:
/home/gitlab-runner/builds/5cd61c66/0/OpenShot/libopenshot/src/../include/ClipBase.h:108:11: error: ‘virtual openshot::ClipBase::~ClipBase()’ cannot be overloaded
   virtual ~ClipBase() = default;
           ^
/home/gitlab-runner/builds/5cd61c66/0/OpenShot/libopenshot/src/../include/ClipBase.h:75:11: error: with ‘virtual openshot::ClipBase::~ClipBase()’
   virtual ~ClipBase();
           ^
In file included from /home/gitlab-runner/builds/5cd61c66/0/OpenShot/libopenshot/src/../include/AudioReaderSource.h:43:0,
                 from /home/gitlab-runner/builds/5cd61c66/0/OpenShot/libopenshot/src/AudioReaderSource.cpp:31:
/home/gitlab-runner/builds/5cd61c66/0/OpenShot/libopenshot/src/../include/ReaderBase.h:156:11: error: ‘virtual openshot::ReaderBase::~ReaderBase()’ cannot be overloaded
   virtual ~ReaderBase() = default;
           ^
/home/gitlab-runner/builds/5cd61c66/0/OpenShot/libopenshot/src/../include/ReaderBase.h:113:14: error: with ‘virtual openshot::ReaderBase::~ReaderBase()’
      virtual ~ReaderBase();
              ^
make[2]: *** [src/CMakeFiles/openshot.dir/AudioReaderSource.cpp.o] Error 1
make[1]: *** [src/CMakeFiles/openshot.dir/all] Error 2
make: *** [all] Error 2
ERROR: Job failed: exit status 1
```

**Comment by jonoomph** [comment]
I'm not sure how Travis succeeded with this one. Hmmm.

**Comment by jonoomph** [comment]
I'm working on a fix. Apparently, we have some duplicate definitions of the destructor, and an implementation of a virtual destructor... which breaks on my builder and older compilers (I guess)

**Comment by jonoomph** [comment]
@sparfenyuk Okay, I think we are fixed up now. Please take a look at #259, and let me know if you see any issues.

**Comment by SuslikV** [comment]
@jonoomph I have an issue with the #259 applied - application crashes on startup (Windows) in `Qt5WebKitWidgets.dll`. The crash caused by removed definition of the destructor of the `ReaderBase()` class. 
The webkit libs is from the `mingw-w64-x86_64-qtwebkit 5.212.0alpha2-6` package.

**Comment by ferdnyc** [comment]
@suslikv

Hmm. First thought... I know you've been building on Appveyor so the answer to this is _probably_ yes, but: Did you do a COMPLETE rebuild of the entire source, after applying #259? If there were old object files around that referenced the removed destructor definition, it's possible those weren't recompiled, and that could cause problems. Like I said, if you're doing CI type builds that start from scratch every time then that's unlikely, but I just wanted to rule out that possibility.

Beyond that, how do you know it was the removal of the destructor declaration? Can you provide an error message/screenshot/etc?

**Comment by SuslikV** [comment]
@ferdnyc It is some hard to me to debug this but, what I have is logs and corresponding artifacts (see below)
When I updated my building branch (of libopenshot) on top of the current _develop_ I got crash during application startup (I simply copied and overwritten the libopenshot.dll from the previous build), then I start to back in history until I found that last working _develop_ for me was f11acef193d3d6bf5f2513858eaaa39b492ff6e6 the few next commits included number of changes, so I start to add it manually, one by one until I found that by removing existing destructor definitions of the ReaderBase causes the crash at startup, the removed definition of the ClipBase causes application crash during video file drag-n-drop into OpenShot.

Normal one, user defined destructor definitions https://ci.appveyor.com/project/SuslikV/libopenshot/builds/25695832

-----

_ReaderBase_ default destructor definition (crash on application startup) https://ci.appveyor.com/project/SuslikV/libopenshot/builds/25697428/artifacts
(screenshot below the spoiler)

<details>
  <summary>Click to expand</summary>

OpenShot crash in Qt5WebKitWidgets dll (default destructor defined by compiler c++11) 01

</details>

-----

_ClipBase_ default destructor definition (crash on file drop) https://ci.appveyor.com/project/SuslikV/libopenshot/builds/25698372
(screenshot below the spoiler)

<details>
  <summary>Click to expand</summary>

OpenShot crash in ntdll dll (default destructor defined by compiler c++11) 02

</details>

-----

The frozen application were I made my test is inside the artifacts of https://ci.appveyor.com/project/SuslikV/openshot-qt/builds/24819868/artifacts

-----

Maybe there should be some additional key to the compiler that I missed? Maybe my compiler can't add destructors definitions automatically? Just adding _default_ doesn't causes error but removing existing definitions - does. It makes for me quite opposite change - the bulletproof user defined destructor was replaces by compiler unknown behavior... Were I did wrong? May you help me? Any other thoughts?

**Comment by SuslikV** [comment]
I found the solution. I totally forgot the Frozen nature of the OpenShot.
The `lib\_openshot.pyd` should be updated together with the `libopenshot.dll` when you overwriting the libs in your setup. The building destination of `.pyd` was `lib\site-packages` , I just forgot about it and copied only `.dll` file. Problem is gone now (I copied it both).

**Comment by ferdnyc** [comment]
I figured it was some sort of object-mismatch issue. A deleted destructor shouldn't manifest as a runtime error. It would either be a compile error or nothing, since once it's deleted, there shouldn't be anything to reference it anymore.

Glad you figured it out!

**Comment by ferdnyc** [comment]
image

I'm more concerned about this, now. "Create and edit stunning videos, movies, and animations has stopped working"? Is that since we started using Verpatch? Because, we're clearly setting the wrong strings into the wrong key, if that's how the crash dialog comes up.

**Comment by ferdnyc** [comment]
Wait, I forgot, Verpatch is only in the GitLab builders for the official releases, it's not normally part of user builds. So, _you're_ clearly setting the wrong string into the wrong key, in your own builds. :wink:

**Comment by ferdnyc** [comment]
(But if you're using the same verpatch command line I used in my PR, then I may have made the same mistake with the official builds. I should probably check that.)

**Comment by SuslikV** [comment]
@ferdnyc  The screenshots are from the pre-version-info commits. I din't set any new strings. Only the version info string was converted all into the numbers by me personally. And of course, I didn't use `verpatch` tool. The job is done in cx_Freeze only. The _targetName_ was not set, so it is still `launch.exe` file.

**Comment by ferdnyc** [comment]
@SuslikV 

My mistake, sorry. How very odd! I wonder where it's even _getting_ that string from?

I've definitely crashed OpenShot in the past, and I don't remember ever seeing a message like that.

**Comment by SuslikV** [comment]
@ferdnyc You can do it every time you wish. Make a project file to crash OpenShot application and open it. For example, the one is based on https://github.com/OpenShot/openshot-qt/issues/2854 report:

The `crash-emulation-01.osp` contents
```
{
    "effects": [
        {
            "type": "Mask",
            "reader": {
                "type": "QtImageReader"
            }
        }
    ]
}
```
It crashes for me in `libstdc++-6.dll`, Windows 7 x64.

**Pull request merged**